File: server/node/crypto.js
const { properties } = require("./properties");

const CryptoJS = require("crypto-js");

const appPassword = properties.appPassword;

function CryptoJSAesEncrypt(passphrase, plainText) {
  const salt = CryptoJS.lib.WordArray.random(256);
  const iv = CryptoJS.lib.WordArray.random(16);

  const key = CryptoJS.PBKDF2(passphrase, salt, { hasher: CryptoJS.algo.SHA512, keySize: 64 / 8, iterations: 999 });

  const encrypted = CryptoJS.AES.encrypt(plainText, key, { iv });

  return JSON.stringify({
    ciphertext: CryptoJS.enc.Base64.stringify(encrypted.ciphertext),
    salt: CryptoJS.enc.Hex.stringify(salt),
    iv: CryptoJS.enc.Hex.stringify(iv),
  });
}

function CryptoJSAesDecrypt(passphrase, encryptedJsonString) {
  const jsondata = JSON.parse(encryptedJsonString);

  const salt = CryptoJS.enc.Hex.parse(jsondata.salt);
  const iv = CryptoJS.enc.Hex.parse(jsondata.iv);
  const key = CryptoJS.PBKDF2(passphrase, salt, { hasher: CryptoJS.algo.SHA512, keySize: 64 / 8, iterations: 999 });

  const decrypted = CryptoJS.AES.decrypt(jsondata.ciphertext, key, { iv });

  return decrypted.toString(CryptoJS.enc.Utf8);
}

// Same PBKDF2-HMAC-SHA512 (999 iterations) + AES-256-CBC scheme as server/php/crypto.php,
// server/java's CryptoService, and server/python's crypto.py — matches the frontend's
// encryptDataPhp/decryptDataPhp in web-app/src/utils/crypto.ts.
exports.encryptData = function (plainTextString) {
  return Buffer.from(CryptoJSAesEncrypt(appPassword, plainTextString)).toString("base64");
};

exports.decryptData = function (base64EnvelopeString) {
  return CryptoJSAesDecrypt(appPassword, Buffer.from(base64EnvelopeString, "base64").toString());
};

File: server/python/app/crypto.py
"""Reimplements the CryptoJSAesEncrypt/CryptoJSAesDecrypt pair from server/php/crypto.php
(also reimplemented in server/java's CryptoService): PBKDF2-HMAC-SHA512 (999 iterations,
256-bit key) + AES-256-CBC, with a random 256-byte salt and 16-byte IV carried alongside the
ciphertext, the whole envelope base64-encoded once more.
"""

import base64
import json
import os

from cryptography.hazmat.primitives import hashes, padding
from cryptography.hazmat.primitives.ciphers import Cipher, algorithms, modes
from cryptography.hazmat.primitives.kdf.pbkdf2 import PBKDF2HMAC

from app import config

_PBKDF2_ITERATIONS = 999
_KEY_LENGTH_BYTES = 32  # 256 bits
_SALT_LENGTH_BYTES = 256
_IV_LENGTH_BYTES = 16


def _derive_key(passphrase: str, salt: bytes) -> bytes:
    kdf = PBKDF2HMAC(
        algorithm=hashes.SHA512(),
        length=_KEY_LENGTH_BYTES,
        salt=salt,
        iterations=_PBKDF2_ITERATIONS,
    )
    return kdf.derive(passphrase.encode("utf-8"))


def encrypt(plain_text: str, passphrase: str = config.APP_PASSWORD) -> str:
    salt = os.urandom(_SALT_LENGTH_BYTES)
    iv = os.urandom(_IV_LENGTH_BYTES)
    key = _derive_key(passphrase, salt)

    padder = padding.PKCS7(algorithms.AES.block_size).padder()
    padded = padder.update(plain_text.encode("utf-8")) + padder.finalize()

    cipher = Cipher(algorithms.AES(key), modes.CBC(iv))
    encryptor = cipher.encryptor()
    ciphertext = encryptor.update(padded) + encryptor.finalize()

    envelope = {
        "ciphertext": base64.b64encode(ciphertext).decode("ascii"),
        "iv": iv.hex(),
        "salt": salt.hex(),
    }
    envelope_json = json.dumps(envelope)
    return base64.b64encode(envelope_json.encode("utf-8")).decode("ascii")


def decrypt(base64_envelope: str, passphrase: str = config.APP_PASSWORD) -> str:
    envelope_json = base64.b64decode(base64_envelope)
    envelope = json.loads(envelope_json)

    salt = bytes.fromhex(envelope["salt"])
    iv = bytes.fromhex(envelope["iv"])
    ciphertext = base64.b64decode(envelope["ciphertext"])

    key = _derive_key(passphrase, salt)

    cipher = Cipher(algorithms.AES(key), modes.CBC(iv))
    decryptor = cipher.decryptor()
    padded = decryptor.update(ciphertext) + decryptor.finalize()

    unpadder = padding.PKCS7(algorithms.AES.block_size).unpadder()
    plain = unpadder.update(padded) + unpadder.finalize()
    return plain.decode("utf-8")
